pkg/graph/resolvers/user_view_customization.go

```diff
@@ -0,0 +1,156 @@
+func UserViewCustomizationGetByUserID(
+	logger *zap.Logger,
+	store *storage.Store,
+	principal authentication.Principal,
+) (*models.UserViewCustomization, error) {
+	retVal, err := sqlutils.WithTransaction[models.UserViewCustomization](
+		store,
+		func(tx *sqlx.Tx) (*models.UserViewCustomization, error) {
+			uvc, err := storage.UserViewCustomizationGetByUserID(tx, principal.Account().ID)
+			if err != nil {
+
+				// If the user view customization does not exist, create it
+				if errors.Is(err, sql.ErrNoRows) {
+					createdUVC, creationErr := createAndInitializeUserViewCustomization(
+						tx,
+						principal,
+						store,
+						logger,
+					)
+					if creationErr != nil {
+						return nil, creationErr
+					}
+
+					// Assign the uvc value to avoid shadowing
+					uvc = createdUVC
+				} else {
+					return nil, err
+				}
+			}
+
+			return uvc, nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return retVal, nil
+}
+
+func createAndInitializeUserViewCustomization(
+	tx *sqlx.Tx,
+	principal authentication.Principal,
+	store *storage.Store,
+	logger *zap.Logger,
+) (*models.UserViewCustomization, error) {
+	uvcToCreate := models.UserViewCustomization{
+		ViewCustomization:            pq.StringArray{},
+		PossibleOperationalSolutions: pq.StringArray{},
+	}
+
+	// should be created for (and by) the user who made the query
+	uvcToCreate.UserID = principal.Account().ID
+	uvcToCreate.CreatedBy = principal.Account().ID
+
+	customizations, err := getDefaultViewCustomizationsByRole(principal, store)
+	if err != nil {
+		return nil, err
+	}
+
+	uvcToCreate.ViewCustomization = models.ConvertEnumsToStringArray(customizations)
+
+	creationErr := BaseStructPreCreate(logger, &uvcToCreate, principal, store, false)
+	if creationErr != nil {
+		return nil, creationErr
+	}
+
+	return storage.UserViewCustomizationCreate(tx, &uvcToCreate)
+}
+
+// getDefaultViewCustomizationsByRole returns a default view customization slice for a principal based on their role
+//
+// Note: The order of these role checks is important. A user who, for example, has AllowAssessment() will ALSO have AllowUser()
+// Therefore, we should check AllowUser() last
+// See pkg/okta/authentication_middleware.go newPrincipal() for details on how role assignment works
+func getDefaultViewCustomizationsByRole(
+	principal authentication.Principal,
+	store *storage.Store,
+) ([]models.ViewCustomizationType, error) {
+	var customizations []models.ViewCustomizationType
+
+	if principal.AllowASSESSMENT() {
+		// Assessment users should have 1 or 2 customizations by default:
+		// 1. My Model Plans (Only present for Assessment users who are collaborators on at least one model plan)
+		// 2. All Model Plans (Always present for Assessment users)
+		userCollaborationsCount, collabCountErr := store.PlanCollaboratorGetCountByUserID(principal.Account().ID)
```

Comment:
(this might be a paradigm we want to revisit, it's the anonymous struct that results in a nil account). It would be nice if we could guarantee an Account exists when this method is called.

Reply:
3d663869